File: packages/ccode/src/autonomous/safety/constraints.ts
import { Log } from "@/util/log"
import { Bus } from "@/bus"
import { AutonomousEvent } from "../events"
import z from "zod"
import {
  createSafetyGuard,
  type SafetyGuardHandleType,
  type NapiResourceBudget,
  type NapiResourceUsage,
  type NapiConstraintCheckResult,
  type NapiResourceWarning,
} from "@codecoder-ai/core"

const log = Log.create({ service: "autonomous.safety.constraints" })

/**
 * Resource budget limits
 */
export interface ResourceBudget {
  maxTokens: number
  maxCostUSD: number
  maxDurationMinutes: number
  maxFilesChanged: number
  maxActions: number
}

/**
 * Current resource usage
 */
export interface ResourceUsage {
  tokensUsed: number
  costUSD: number
  durationMinutes: number
  filesChanged: number
  actionsPerformed: number
}

/**
 * Safety check result
 */
export interface SafetyCheckResult {
  safe: boolean
  reason?: string
  resource?: keyof ResourceBudget
  current?: number
  limit?: number
}

/**
 * Safety configuration
 */
export interface SafetyConfig {
  budget: ResourceBudget
  warnThreshold: number // Percentage (0-100) to warn at
  hardLimit: boolean // Enforce hard limits or just warn
}

/**
 * Default resource budget
 */
export const DEFAULT_BUDGET: ResourceBudget = {
  maxTokens: 1_000_000,
  maxCostUSD: 10.0,
  maxDurationMinutes: 30,
  maxFilesChanged: 50,
  maxActions: 100,
}

/**
 * Map resource type string to ResourceBudget key
 */
function mapResourceToKey(resource: string | null | undefined): keyof ResourceBudget | undefined {
  switch (resource) {
    case "tokens":
      return "maxTokens"
    case "cost":
      return "maxCostUSD"
    case "time":
    case "duration":
      return "maxDurationMinutes"
    case "files":
      return "maxFilesChanged"
    case "actions":
      return "maxActions"
    default:
      return undefined
  }
}

/**
 * Safety guard for resource limits - Rust-backed implementation
 *
 * Monitors and enforces resource constraints using native implementation.
 * Event publishing is handled in TypeScript for Bus compatibility.
 */
export class SafetyGuard {
  private handle: SafetyGuardHandleType
  private config: SafetyConfig
  private sessionId: string
  private warningSent: Set<keyof ResourceBudget> = new Set()

  constructor(sessionId: string, budget?: Partial<ResourceBudget>) {
    this.sessionId = sessionId
    const fullBudget = { ...DEFAULT_BUDGET, ...budget }
    this.config = {
      budget: fullBudget,
      warnThreshold: 80,
      hardLimit: true,
    }

    // Convert to NAPI format
    const napiBudget: NapiResourceBudget = {
      maxTokens: fullBudget.maxTokens,
      maxCostUsd: fullBudget.maxCostUSD,
      maxDurationMinutes: fullBudget.maxDurationMinutes,
      maxFilesChanged: fullBudget.maxFilesChanged,
      maxActions: fullBudget.maxActions,
    }

    // Create Rust-backed handle
    const handle = createSafetyGuard?.(sessionId, napiBudget)
    if (!handle) {
      throw new Error("Native SafetyGuard not available. Build native modules with `cargo build` in services/zero-core.")
    }
    this.handle = handle
  }

  /**
   * Check if an action is safe to perform
   */
  async check(action?: keyof ResourceBudget, additionalCost?: Partial<ResourceUsage>): Promise<SafetyCheckResult> {
    // Convert additional cost to NAPI format
    const napiUsage: NapiResourceUsage | undefined = additionalCost
      ? {
          tokensUsed: additionalCost.tokensUsed,
          costUsd: additionalCost.costUSD,
          durationMinutes: additionalCost.durationMinutes,
          filesChanged: additionalCost.filesChanged,
          actionsPerformed: additionalCost.actionsPerformed,
        }
      : undefined

    // Call Rust check with warnings
    const { result, warnings } = this.handle.check(napiUsage)

    // Publish warnings to Bus (TypeScript layer handles events)
    for (const warning of warnings) {
      this.publishWarning(warning)
    }

    // Convert result
    if (!result.safe && this.config.hardLimit) {
      const resourceKey = mapResourceToKey(result.resource)
      return {
        safe: false,
        reason: result.reason ?? undefined,
        resource: resourceKey,
        current: result.current ?? undefined,
        limit: result.limit ?? undefined,
      }
    }

    return { safe: true }
  }

  /**
   * Publish resource warning to Bus
   */
  private publishWarning(warning: NapiResourceWarning): void {
    const resourceKey = mapResourceToKey(warning.resource)
    if (!resourceKey || this.warningSent.has(resourceKey)) {
      return
    }

    this.warningSent.add(resourceKey)

    log.warn("Resource warning", {
      resource: warning.resource,
      current: warning.current,
      limit: warning.limit,
      percentage: warning.percentage,
    })

    Bus.publish(AutonomousEvent.ResourceWarning, {
      sessionId: this.sessionId,
      resource: warning.resource as "tokens" | "cost" | "time" | "files" | "actions",
      current: warning.current,
      limit: warning.limit,
      percentage: warning.percentage,
    })
  }

  /**
   * Record resource usage
   */
  record(resource: keyof ResourceUsage, value: number): void {
    const usage: NapiResourceUsage = {}
    switch (resource) {
      case "tokensUsed":
        usage.tokensUsed = value
        break
      case "costUSD":
        usage.costUsd = value
        break
      case "durationMinutes":
        usage.durationMinutes = value
        break
      case "filesChanged":
        usage.filesChanged = value
        break
      case "actionsPerformed":
        usage.actionsPerformed = value
        break
    }
    this.handle.record(usage)
  }

  /**
   * Get current usage
   */
  getCurrentUsage(): ResourceUsage {
    const usage = this.handle.getUsage()
    return {
      tokensUsed: usage.tokensUsed ?? 0,
      costUSD: usage.costUsd ?? 0,
      durationMinutes: usage.durationMinutes ?? 0,
      filesChanged: usage.filesChanged ?? 0,
      actionsPerformed: usage.actionsPerformed ?? 0,
    }
  }

  /**
   * Get remaining budget
   */
  getRemaining(): Partial<ResourceBudget> {
    const remaining = this.handle.getRemaining()
    return {
      maxTokens: remaining.maxTokens ?? 0,
      maxCostUSD: remaining.maxCostUsd ?? 0,
      maxDurationMinutes: remaining.maxDurationMinutes ?? 0,
      maxFilesChanged: remaining.maxFilesChanged ?? 0,
      maxActions: remaining.maxActions ?? 0,
    }
  }

  /**
   * Get surplus ratio (0-1)
   */
  getSurplusRatio(): number {
    return this.handle.getSurplusRatio()
  }

  /**
   * Get warnings count (number of resources that have triggered warnings)
   */
  getWarningsCount(): number {
    return this.warningSent.size
  }

  /**
   * Reset usage
   */
  reset(): void {
    this.handle.reset()
    this.warningSent.clear()
  }

  /**
   * Update budget
   */
  updateBudget(budget: Partial<ResourceBudget>): void {
    this.config.budget = { ...this.config.budget, ...budget }

    const napiBudget: NapiResourceBudget = {
      maxTokens: budget.maxTokens,
      maxCostUsd: budget.maxCostUSD,
      maxDurationMinutes: budget.maxDurationMinutes,
      maxFilesChanged: budget.maxFilesChanged,
      maxActions: budget.maxActions,
    }
    this.handle.updateBudget(napiBudget)
  }

  /**
   * Get config
   */
  getConfig(): SafetyConfig {
    return { ...this.config }
  }

  /**
   * Serialize guard state
   */
  serialize(): { usage: ResourceUsage; startTime: number; config: SafetyConfig } {
    const json = this.handle.serialize()
    try {
      const data = JSON.parse(json)
      return {
        usage: this.getCurrentUsage(),
        startTime: data.startTime ?? Date.now(),
        config: this.config,
      }
    } catch {
      return {
        usage: this.getCurrentUsage(),
        startTime: Date.now(),
        config: this.config,
      }
    }
  }

  /**
   * Restore from serialized data
   */
  static deserialize(
    data: {
      usage: ResourceUsage
      startTime: number
      config: SafetyConfig
    },
    sessionId: string,
  ): SafetyGuard {
    const guard = new SafetyGuard(sessionId, data.config.budget)

    // Add recorded usage to the new guard
    if (data.usage.tokensUsed > 0) {
      guard.handle.addTokens(data.usage.tokensUsed)
    }
    if (data.usage.costUSD > 0) {
      guard.handle.addCost(data.usage.costUSD)
    }
    if (data.usage.filesChanged > 0) {
      guard.handle.addFilesChanged(data.usage.filesChanged)
    }
    if (data.usage.actionsPerformed > 0) {
      guard.handle.addActions(data.usage.actionsPerformed)
    }

    return guard
  }
}

/**
 * Schema for resource budget configuration
 */
export const ResourceBudgetSchema = z.object({
  maxTokens: z.number().min(0).optional(),
  maxCostUSD: z.number().min(0).optional(),
  maxDurationMinutes: z.number().min(0).optional(),
  maxFilesChanged: z.number().min(0).optional(),
  maxActions: z.number().min(0).optional(),
})

/**
 * Parse resource budget from config
 */
export function parseResourceBudget(config: unknown): ResourceBudget {
  const parsed = ResourceBudgetSchema.partial().parse(config)
  return {
    ...DEFAULT_BUDGET,
    ...parsed,
  }
}

File: services/zero-core/src/autonomous/constraints.rs
//! Safety Constraints for Resource Budget Management
//!
//! Monitors and enforces resource constraints during autonomous execution:
//! - Token usage limits
//! - Cost limits (USD)
//! - Duration limits
//! - File change limits
//! - Action limits
//!
//! # Philosophy
//!
//! Following the "可用余量" (Surplus) principle, constraints ensure we maintain
//! sufficient margin to recover from errors and complete objectives.

use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::time::Instant;

// ══════════════════════════════════════════════════════════════════════════════
// Types
// ══════════════════════════════════════════════════════════════════════════════

/// Resource budget limits
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceBudget {
    /// Maximum tokens to use
    pub max_tokens: usize,
    /// Maximum cost in USD
    pub max_cost_usd: f64,
    /// Maximum duration in minutes
    pub max_duration_minutes: f64,
    /// Maximum files changed
    pub max_files_changed: usize,
    /// Maximum actions performed
    pub max_actions: usize,
}

impl Default for ResourceBudget {
    fn default() -> Self {
        Self {
            max_tokens: 1_000_000,
            max_cost_usd: 10.0,
            max_duration_minutes: 30.0,
            max_files_changed: 50,
            max_actions: 100,
        }
    }
}

/// Current resource usage
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceUsage {
    /// Tokens used so far
    pub tokens_used: usize,
    /// Cost in USD so far
    pub cost_usd: f64,
    /// Duration in minutes
    pub duration_minutes: f64,
    /// Files changed
    pub files_changed: usize,
    /// Actions performed
    pub actions_performed: usize,
}

impl ResourceUsage {
    /// Add usage from another ResourceUsage
    pub fn add(&mut self, other: &ResourceUsage) {
        self.tokens_used += other.tokens_used;
        self.cost_usd += other.cost_usd;
        self.files_changed += other.files_changed;
        self.actions_performed += other.actions_performed;
        // Note: duration_minutes is calculated from start time, not added
    }
}

/// Resource type that exceeded limit
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ResourceType {
    Tokens,
    Cost,
    Duration,
    Files,
    Actions,
}

impl std::fmt::Display for ResourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Tokens => write!(f, "tokens"),
            Self::Cost => write!(f, "cost"),
            Self::Duration => write!(f, "time"),
            Self::Files => write!(f, "files"),
            Self::Actions => write!(f, "actions"),
        }
    }
}

/// Safety check result for resource constraints
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConstraintCheckResult {
    /// Whether the check passed (safe to continue)
    pub safe: bool,
    /// Reason for failure (if not safe)
    pub reason: Option<String>,
    /// Resource type that exceeded limit
    pub resource: Option<ResourceType>,
    /// Current usage value
    pub current: Option<f64>,
    /// Limit value
    pub limit: Option<f64>,
}

impl ConstraintCheckResult {
    /// Create a passing result
    pub fn ok() -> Self {
        Self {
            safe: true,
            reason: None,
            resource: None,
            current: None,
            limit: None,
        }
    }

    /// Create a failing result
    pub fn fail(resource: ResourceType, current: f64, limit: f64) -> Self {
        Self {
            safe: false,
            reason: Some(format!(
                "Resource limit exceeded for {}: {:.2}/{:.2}",
                resource, current, limit
            )),
            resource: Some(resource),
            current: Some(current),
            limit: Some(limit),
        }
    }
}

/// Resource warning (approaching limit)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceWarning {
    pub session_id: String,
    pub resource: ResourceType,
    pub current: f64,
    pub limit: f64,
    pub percentage: u32,
}

/// Safety guard configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SafetyConfig {
    /// Resource budget
    pub budget: ResourceBudget,
    /// Percentage (0-100) at which to warn
    pub warn_threshold: u32,
    /// Whether to enforce hard limits
    pub hard_limit: bool,
}

impl Default for SafetyConfig {
    fn default() -> Self {
        Self {
            budget: ResourceBudget::default(),
            warn_threshold: 80,
            hard_limit: true,
        }
    }
}

// ══════════════════════════════════════════════════════════════════════════════
// SafetyGuard
// ══════════════════════════════════════════════════════════════════════════════

/// Safety guard for monitoring and enforcing resource constraints
#[derive(Debug)]
pub struct SafetyGuard {
    session_id: String,
    config: SafetyConfig,
    usage: ResourceUsage,
    start_time: Instant,
    warnings_sent: HashSet<ResourceType>,
}

impl SafetyGuard {
    /// Create a new safety guard
    pub fn new(session_id: impl Into<String>, budget: ResourceBudget) -> Self {
        Self {
            session_id: session_id.into(),
            config: SafetyConfig {
                budget,
                ..Default::default()
            },
            usage: ResourceUsage::default(),
            start_time: Instant::now(),
            warnings_sent: HashSet::new(),
        }
    }

    /// Create with default budget
    pub fn with_defaults(session_id: impl Into<String>) -> Self {
        Self::new(session_id, ResourceBudget::default())
    }

    /// Create with full configuration
    pub fn with_config(session_id: impl Into<String>, config: SafetyConfig) -> Self {
        Self {
            session_id: session_id.into(),
            config,
            usage: ResourceUsage::default(),
            start_time: Instant::now(),
            warnings_sent: HashSet::new(),
        }
    }

    /// Get session ID
    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    /// Get configuration
    pub fn config(&self) -> &SafetyConfig {
        &self.config
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Resource Checking
    // ─────────────────────────────────────────────────────────────────────────

    /// Check if an operation is safe to perform
    ///
    /// Optionally accepts additional cost to add before checking.
    /// Returns warnings if approaching limits.
    pub fn check(&mut self, additional: Option<&ResourceUsage>) -> (ConstraintCheckResult, Vec<ResourceWarning>) {
        // Update duration
        self.usage.duration_minutes = self.start_time.elapsed().as_secs_f64() / 60.0;

        // Add additional costs
        if let Some(cost) = additional {
            self.usage.add(cost);
        }

        let mut warnings = Vec::new();
        let budget = &self.config.budget;

        // Check each resource
        let checks = [
            (
                ResourceType::Tokens,
                self.usage.tokens_used as f64,
                budget.max_tokens as f64,
            ),
            (
                ResourceType::Cost,
                self.usage.cost_usd,
                budget.max_cost_usd,
            ),
            (
                ResourceType::Duration,
                self.usage.duration_minutes,
                budget.max_duration_minutes,
            ),
            (
                ResourceType::Files,
                self.usage.files_changed as f64,
                budget.max_files_changed as f64,
            ),
            (
                ResourceType::Actions,
                self.usage.actions_performed as f64,
                budget.max_actions as f64,
            ),
        ];

        for (resource, current, limit) in checks {
            // Check hard limit
            if current >= limit && self.config.hard_limit {
                return (ConstraintCheckResult::fail(resource, current, limit), warnings);
            }

            // Check warning threshold
            let percentage = ((current / limit) * 100.0) as u32;
            if percentage >= self.config.warn_threshold && !self.warnings_sent.contains(&resource) {
                self.warnings_sent.insert(resource);
                warnings.push(ResourceWarning {
                    session_id: self.session_id.clone(),
                    resource,
                    current,
                    limit,
                    percentage,
                });
            }
        }

        (ConstraintCheckResult::ok(), warnings)
    }

    /// Quick check without warnings (does not modify state)
    pub fn quick_check(&self) -> ConstraintCheckResult {
        let budget = &self.config.budget;
        let duration = self.start_time.elapsed().as_secs_f64() / 60.0;

        let checks = [
            (
                ResourceType::Tokens,
                self.usage.tokens_used as f64,
                budget.max_tokens as f64,
            ),
            (
                ResourceType::Cost,
                self.usage.cost_usd,
                budget.max_cost_usd,
            ),
            (ResourceType::Duration, duration, budget.max_duration_minutes),
            (
                ResourceType::Files,
                self.usage.files_changed as f64,
                budget.max_files_changed as f64,
            ),
            (
                ResourceType::Actions,
                self.usage.actions_performed as f64,
                budget.max_actions as f64,
            ),
        ];

        for (resource, current, limit) in checks {
            if current >= limit && self.config.hard_limit {
                return ConstraintCheckResult::fail(resource, current, limit);
            }
        }

        ConstraintCheckResult::ok()
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Usage Tracking
    // ─────────────────────────────────────────────────────────────────────────

    /// Record resource usage
    pub fn record(&mut self, usage: &ResourceUsage) {
        self.usage.add(usage);
    }

    /// Add tokens used
    pub fn add_tokens(&mut self, tokens: usize) {
        self.usage.tokens_used += tokens;
    }

    /// Add cost
    pub fn add_cost(&mut self, cost: f64) {
        self.usage.cost_usd += cost;
    }

    /// Add files changed
    pub fn add_files_changed(&mut self, count: usize) {
        self.usage.files_changed += count;
    }

    /// Add actions performed
    pub fn add_actions(&mut self, count: usize) {
        self.usage.actions_performed += count;
    }

    /// Get current usage
    pub fn get_usage(&self) -> ResourceUsage {
        let mut usage = self.usage.clone();
        usage.duration_minutes = self.start_time.elapsed().as_secs_f64() / 60.0;
        usage
    }

    /// Get remaining budget
    pub fn get_remaining(&self) -> ResourceBudget {
        let usage = self.get_usage();
        let budget = &self.config.budget;

        ResourceBudget {
            max_tokens: budget.max_tokens.saturating_sub(usage.tokens_used),
            max_cost_usd: (budget.max_cost_usd - usage.cost_usd).max(0.0),
            max_duration_minutes: (budget.max_duration_minutes - usage.duration_minutes).max(0.0),
            max_files_changed: budget.max_files_changed.saturating_sub(usage.files_changed),
            max_actions: budget.max_actions.saturating_sub(usage.actions_performed),
        }
    }

    /// Get surplus ratio (0.0-1.0, higher = more resources remaining)
    pub fn get_surplus_ratio(&self) -> f64 {
        let remaining = self.get_remaining();
        let budget = &self.config.budget;

        let ratios = [
            remaining.max_tokens as f64 / budget.max_tokens as f64,
            remaining.max_cost_usd / budget.max_cost_usd,
            remaining.max_duration_minutes / budget.max_duration_minutes,
            remaining.max_files_changed as f64 / budget.max_files_changed as f64,
            remaining.max_actions as f64 / budget.max_actions as f64,
        ];

        ratios.iter().sum::<f64>() / ratios.len() as f64
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Configuration
    // ─────────────────────────────────────────────────────────────────────────

    /// Update budget
    pub fn update_budget(&mut self, budget: ResourceBudget) {
        self.config.budget = budget;
    }

    /// Reset usage (restart tracking)
    pub fn reset(&mut self) {
        self.usage = ResourceUsage::default();
        self.start_time = Instant::now();
        self.warnings_sent.clear();
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Serialization
    // ─────────────────────────────────────────────────────────────────────────

    /// Serialize current state (for persistence/restore)
    pub fn serialize(&self) -> serde_json::Value {
        let usage = self.get_usage();
        serde_json::json!({
            "sessionId": self.session_id,
            "usage": usage,
            "config": self.config,
            "startTime": chrono::Utc::now().timestamp_millis()
        })
    }
}

// ══════════════════════════════════════════════════════════════════════════════
// Tests
// ══════════════════════════════════════════════════════════════════════════════

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_safety_guard_default() {
        let guard = SafetyGuard::with_defaults("test-session");
        assert_eq!(guard.session_id(), "test-session");
        assert!(guard.quick_check().safe);
    }

    #[test]
    fn test_usage_tracking() {
        let mut guard = SafetyGuard::with_defaults("test");

        guard.add_tokens(1000);
        guard.add_cost(0.5);
        guard.add_files_changed(1);
        guard.add_actions(2);

        let usage = guard.get_usage();
        assert_eq!(usage.tokens_used, 1000);
        assert!((usage.cost_usd - 0.5).abs() < 0.001);
        assert_eq!(usage.files_changed, 1);
        assert_eq!(usage.actions_performed, 2);
    }

    #[test]
    fn test_limit_exceeded() {
        let budget = ResourceBudget {
            max_tokens: 100,
            ..Default::default()
        };
        let mut guard = SafetyGuard::new("test", budget);

        guard.add_tokens(100);
        let result = guard.quick_check();

        assert!(!result.safe);
        assert_eq!(result.resource, Some(ResourceType::Tokens));
    }

    #[test]
    fn test_warning_threshold() {
        let budget = ResourceBudget {
            max_tokens: 100,
            ..Default::default()
        };
        let config = SafetyConfig {
            budget,
            warn_threshold: 80,
            hard_limit: true,
        };
        let mut guard = SafetyGuard::with_config("test", config);

        guard.add_tokens(80);
        let (result, warnings) = guard.check(None);

        assert!(result.safe);
        assert_eq!(warnings.len(), 1);
        assert_eq!(warnings[0].resource, ResourceType::Tokens);
    }

    #[test]
    fn test_remaining_budget() {
        let budget = ResourceBudget {
            max_tokens: 1000,
            max_cost_usd: 10.0,
            ..Default::default()
        };
        let mut guard = SafetyGuard::new("test", budget);

        guard.add_tokens(300);
        guard.add_cost(3.0);

        let remaining = guard.get_remaining();
        assert_eq!(remaining.max_tokens, 700);
        assert!((remaining.max_cost_usd - 7.0).abs() < 0.001);
    }

    #[test]
    fn test_surplus_ratio() {
        let budget = ResourceBudget {
            max_tokens: 1000,
            max_cost_usd: 10.0,
            max_duration_minutes: 30.0,
            max_files_changed: 50,
            max_actions: 100,
        };
        let guard = SafetyGuard::new("test", budget);

        // Fresh guard should have high surplus
        let ratio = guard.get_surplus_ratio();
        assert!(ratio > 0.9);
    }

    #[test]
    fn test_reset() {
        let mut guard = SafetyGuard::with_defaults("test");

        guard.add_tokens(1000);
        guard.add_cost(5.0);

        guard.reset();

        let usage = guard.get_usage();
        assert_eq!(usage.tokens_used, 0);
        assert!((usage.cost_usd).abs() < 0.001);
    }

    #[test]
    fn test_serialization() {
        let mut guard = SafetyGuard::with_defaults("test");
        guard.add_tokens(500);

        let json = guard.serialize();
        assert!(json.get("usage").is_some());
        assert!(json.get("config").is_some());
    }
}
